### Feature Engineering

In [9]:
import pandas as pd
import sys
sys.path.insert(0, '/10 A KAI 2/Week6/Credit_Scoring/')

In [10]:
from Script.Feature_engi import Feature_Engineering

In [11]:
df = pd.read_csv('../Data/data.csv')

In [12]:
fut_eng = Feature_Engineering(df)

### Remove unneccery columns 

In [13]:
attrib=['BatchId',  'SubscriptionId', 'CurrencyCode', 'CountryCode', 'ProductId']
fut_eng.remove_some_cols(attrib)

TransactionId       AccountId       CustomerId    ProviderId  \
0       TransactionId_76871  AccountId_3957  CustomerId_4406  ProviderId_6   
1       TransactionId_73770  AccountId_4841  CustomerId_4406  ProviderId_4   
2       TransactionId_26203  AccountId_4229  CustomerId_4683  ProviderId_6   
3         TransactionId_380   AccountId_648   CustomerId_988  ProviderId_1   
4       TransactionId_28195  AccountId_4841   CustomerId_988  ProviderId_4   
...                     ...             ...              ...           ...   
95657   TransactionId_89881  AccountId_4841  CustomerId_3078  ProviderId_4   
95658   TransactionId_91597  AccountId_3439  CustomerId_3874  ProviderId_6   
95659   TransactionId_82501  AccountId_4841  CustomerId_3874  ProviderId_4   
95660  TransactionId_136354  AccountId_1346  CustomerId_1709  ProviderId_6   
95661   TransactionId_35670  AccountId_4841  CustomerId_1709  ProviderId_4   

          ProductCategory    ChannelId   Amount  Value  TransactionStartTime  \
0                 airtime  ChannelId_3   1000.0   1000  2018-11-15T02:18:49Z   
1      financial_services  ChannelId_2    -20.0     20  2018-11-15T02:19:08Z   
2                 airtime  ChannelId_3    500.0    500  2018-11-15T02:44:21Z   
3            utility_bill  ChannelId_3  20000.0  21800  2018-11-15T03:32:55Z   
4      financial_services  ChannelId_2   -644.0    644  2018-11-15T03:34:21Z   
...                   ...          ...      ...    ...                   ...   
95657  financial_services  ChannelId_2  -1000.0   1000  2019-02-13T09:54:09Z   
95658             airtime  ChannelId_3   1000.0   1000  2019-02-13T09:54:25Z   
95659  financial_services  ChannelId_2    -20.0     20  2019-02-13T09:54:35Z   
95660                  tv  ChannelId_3   3000.0   3000  2019-02-13T10:01:10Z   
95661  financial_services  ChannelId_2    -60.0     60  2019-02-13T10:01:28Z   

       PricingStrategy  FraudResult  
0                    2            0  
1                    2            0  
2                    2            0  
3                    2            0  
4                    2            0  
...                ...          ...  
95657                2            0  
95658                2            0  
95659                2            0  
95660                2            0  
95661                2            0  

[95662 rows x 11 columns]

#### Aggregate Credit and Debit transaction amount in the customer label 
1. **Total Credit Amount**: Summed the values in the Credit column to determine the total credit transactions for each customer.
2. **Total Debit Amount**: Summed the values in the Debit column to determine the total debit transactions for each customer.
3. **Average Transaction Amount**: Calculated the mean of the Value column, which represents the average transaction value for each customer.
4.  **Transaction Count**: Counted the number of transactions per CustomerId using the TransactionId column.
5.  **Total Unique Account Count**: Counted the unique accounts associated with each customer using the AccountId column.
6.  **Standard Deviation of Transaction Amount**: Computed the standard deviation of the Amount column to assess the variability of transaction amounts. The ddof=0 parameter ensures a population standard deviation is calculated, and if there are fewer than two entries, NaN is replaced with zero. 

In [6]:
fut_eng.feature_engineering_transaction()

CustomerId  Total_Credit_Amount  Total_Debit_Amount  \
0        CustomerId_1                  0.0             10000.0   
1       CustomerId_10                  0.0             10000.0   
2     CustomerId_1001              25200.0              5200.0   
3     CustomerId_1002               4500.0               275.0   
4     CustomerId_1003              26000.0              6000.0   
...               ...                  ...                 ...   
3737   CustomerId_992              26000.0              6000.0   
3738   CustomerId_993              26000.0              6000.0   
3739   CustomerId_994             578000.0             34127.0   
3740   CustomerId_996             145000.0              6000.0   
3741   CustomerId_998             131500.0             31500.0   

      Average_Transaction_Amount  Transaction_Count  \
0                   10000.000000                  1   
1                   10000.000000                  1   
2                    6080.000000                  5   
3                     434.090909                 11   
4                    5333.333333                  6   
...                          ...                ...   
3737                 5333.333333                  6   
3738                 6400.000000                  5   
3739                 6079.970297                101   
3740                 8882.352941                 17   
3741                 7409.090909                 22   

      Total_Unique_Account_Count  distinct_product_categories  \
0                              1                            1   
1                              1                            1   
2                              2                            2   
3                              2                            2   
4                              3                            2   
...                          ...                          ...   
3737                           3                            2   
3738                           3                            2   
3739                           2                            3   
3740                           3                            2   
3741                           3                            2   

      Standard_Deviation_Transaction_Amount  
0                                  0.000000  
1                                  0.000000  
2                               5866.515150  
3                                534.414795  
4                               5505.048188  
...                                     ...  
3737                            5557.777334  
3738                            6033.241252  
3739                           14727.203940  
3740                            4300.961508  
3741                            6619.374268  

[3742 rows x 8 columns]

##### The above result shows the aggrigation by customer. totally we have 3742 customer.

In [7]:
df

TransactionId         BatchId       AccountId  \
0       TransactionId_76871   BatchId_36123  AccountId_3957   
1       TransactionId_73770   BatchId_15642  AccountId_4841   
2       TransactionId_26203   BatchId_53941  AccountId_4229   
3         TransactionId_380  BatchId_102363   AccountId_648   
4       TransactionId_28195   BatchId_38780  AccountId_4841   
...                     ...             ...             ...   
95657   TransactionId_89881   BatchId_96668  AccountId_4841   
95658   TransactionId_91597    BatchId_3503  AccountId_3439   
95659   TransactionId_82501  BatchId_118602  AccountId_4841   
95660  TransactionId_136354   BatchId_70924  AccountId_1346   
95661   TransactionId_35670   BatchId_29317  AccountId_4841   

            SubscriptionId       CustomerId CurrencyCode  CountryCode  \
0       SubscriptionId_887  CustomerId_4406          UGX          256   
1      SubscriptionId_3829  CustomerId_4406          UGX          256   
2       SubscriptionId_222  CustomerId_4683          UGX          256   
3      SubscriptionId_2185   CustomerId_988          UGX          256   
4      SubscriptionId_3829   CustomerId_988          UGX          256   
...                    ...              ...          ...          ...   
95657  SubscriptionId_3829  CustomerId_3078          UGX          256   
95658  SubscriptionId_2643  CustomerId_3874          UGX          256   
95659  SubscriptionId_3829  CustomerId_3874          UGX          256   
95660   SubscriptionId_652  CustomerId_1709          UGX          256   
95661  SubscriptionId_3829  CustomerId_1709          UGX          256   

         ProviderId     ProductId     ProductCategory    ChannelId   Amount  \
0      ProviderId_6  ProductId_10             airtime  ChannelId_3   1000.0   
1      ProviderId_4   ProductId_6  financial_services  ChannelId_2    -20.0   
2      ProviderId_6   ProductId_1             airtime  ChannelId_3    500.0   
3      ProviderId_1  ProductId_21        utility_bill  ChannelId_3  20000.0   
4      ProviderId_4   ProductId_6  financial_services  ChannelId_2   -644.0   
...             ...           ...                 ...          ...      ...   
95657  ProviderId_4   ProductId_6  financial_services  ChannelId_2  -1000.0   
95658  ProviderId_6  ProductId_10             airtime  ChannelId_3   1000.0   
95659  ProviderId_4   ProductId_6  financial_services  ChannelId_2    -20.0   
95660  ProviderId_6  ProductId_19                  tv  ChannelId_3   3000.0   
95661  ProviderId_4   ProductId_6  financial_services  ChannelId_2    -60.0   

       Value  TransactionStartTime  PricingStrategy  FraudResult  
0       1000  2018-11-15T02:18:49Z                2            0  
1         20  2018-11-15T02:19:08Z                2            0  
2        500  2018-11-15T02:44:21Z                2            0  
3      21800  2018-11-15T03:32:55Z                2            0  
4        644  2018-11-15T03:34:21Z                2            0  
...      ...                   ...              ...          ...  
95657   1000  2019-02-13T09:54:09Z                2            0  
95658   1000  2019-02-13T09:54:25Z                2            0  
95659     20  2019-02-13T09:54:35Z                2            0  
95660   3000  2019-02-13T10:01:10Z                2            0  
95661     60  2019-02-13T10:01:28Z                2            0  

[95662 rows x 16 columns]

#### Extract Transaction day, year,  month and Hour fron trsnsaction date 

In [14]:
fut_eng.extract_transaction_features()

TransactionId       AccountId       CustomerId    ProviderId  \
0       TransactionId_76871  AccountId_3957  CustomerId_4406  ProviderId_6   
1       TransactionId_73770  AccountId_4841  CustomerId_4406  ProviderId_4   
2       TransactionId_26203  AccountId_4229  CustomerId_4683  ProviderId_6   
3         TransactionId_380   AccountId_648   CustomerId_988  ProviderId_1   
4       TransactionId_28195  AccountId_4841   CustomerId_988  ProviderId_4   
...                     ...             ...              ...           ...   
95657   TransactionId_89881  AccountId_4841  CustomerId_3078  ProviderId_4   
95658   TransactionId_91597  AccountId_3439  CustomerId_3874  ProviderId_6   
95659   TransactionId_82501  AccountId_4841  CustomerId_3874  ProviderId_4   
95660  TransactionId_136354  AccountId_1346  CustomerId_1709  ProviderId_6   
95661   TransactionId_35670  AccountId_4841  CustomerId_1709  ProviderId_4   

          ProductCategory    ChannelId   Amount  Value  \
0                 airtime  ChannelId_3   1000.0   1000   
1      financial_services  ChannelId_2    -20.0     20   
2                 airtime  ChannelId_3    500.0    500   
3            utility_bill  ChannelId_3  20000.0  21800   
4      financial_services  ChannelId_2   -644.0    644   
...                   ...          ...      ...    ...   
95657  financial_services  ChannelId_2  -1000.0   1000   
95658             airtime  ChannelId_3   1000.0   1000   
95659  financial_services  ChannelId_2    -20.0     20   
95660                  tv  ChannelId_3   3000.0   3000   
95661  financial_services  ChannelId_2    -60.0     60   

           TransactionStartTime  PricingStrategy  FraudResult  \
0     2018-11-15 02:18:49+00:00                2            0   
1     2018-11-15 02:19:08+00:00                2            0   
2     2018-11-15 02:44:21+00:00                2            0   
3     2018-11-15 03:32:55+00:00                2            0   
4     2018-11-15 03:34:21+00:00                2            0   
...                         ...              ...          ...   
95657 2019-02-13 09:54:09+00:00                2            0   
95658 2019-02-13 09:54:25+00:00                2            0   
95659 2019-02-13 09:54:35+00:00                2            0   
95660 2019-02-13 10:01:10+00:00                2            0   
95661 2019-02-13 10:01:28+00:00                2            0   

       Transaction_Hour  Transaction_Day  Transaction_Month  Transaction_Year  
0                     2               15                 11              2018  
1                     2               15                 11              2018  
2                     2               15                 11              2018  
3                     3               15                 11              2018  
4                     3               15                 11              2018  
...                 ...              ...                ...               ...  
95657                 9               13                  2              2019  
95658                 9               13                  2              2019  
95659                 9               13                  2              2019  
95660                10               13                  2              2019  
95661                10               13                  2              2019  

[95662 rows x 15 columns]

#### Label encoding for column ProductCategory, ProviderId and ChannelId

In [15]:
column = ['ProductCategory',  'ChannelId', 'ProviderId']
fut_eng.label_encoding_selected_attribute(column)

TransactionId       AccountId       CustomerId  ProviderId  \
0       TransactionId_76871  AccountId_3957  CustomerId_4406           5   
1       TransactionId_73770  AccountId_4841  CustomerId_4406           3   
2       TransactionId_26203  AccountId_4229  CustomerId_4683           5   
3         TransactionId_380   AccountId_648   CustomerId_988           0   
4       TransactionId_28195  AccountId_4841   CustomerId_988           3   
...                     ...             ...              ...         ...   
95657   TransactionId_89881  AccountId_4841  CustomerId_3078           3   
95658   TransactionId_91597  AccountId_3439  CustomerId_3874           5   
95659   TransactionId_82501  AccountId_4841  CustomerId_3874           3   
95660  TransactionId_136354  AccountId_1346  CustomerId_1709           5   
95661   TransactionId_35670  AccountId_4841  CustomerId_1709           3   

       ProductCategory  ChannelId   Amount  Value      TransactionStartTime  \
0                    0          2   1000.0   1000 2018-11-15 02:18:49+00:00   
1                    2          1    -20.0     20 2018-11-15 02:19:08+00:00   
2                    0          2    500.0    500 2018-11-15 02:44:21+00:00   
3                    8          2  20000.0  21800 2018-11-15 03:32:55+00:00   
4                    2          1   -644.0    644 2018-11-15 03:34:21+00:00   
...                ...        ...      ...    ...                       ...   
95657                2          1  -1000.0   1000 2019-02-13 09:54:09+00:00   
95658                0          2   1000.0   1000 2019-02-13 09:54:25+00:00   
95659                2          1    -20.0     20 2019-02-13 09:54:35+00:00   
95660                7          2   3000.0   3000 2019-02-13 10:01:10+00:00   
95661                2          1    -60.0     60 2019-02-13 10:01:28+00:00   

       PricingStrategy  FraudResult  Transaction_Hour  Transaction_Day  \
0                    2            0                 2               15   
1                    2            0                 2               15   
2                    2            0                 2               15   
3                    2            0                 3               15   
4                    2            0                 3               15   
...                ...          ...               ...              ...   
95657                2            0                 9               13   
95658                2            0                 9               13   
95659                2            0                 9               13   
95660                2            0                10               13   
95661                2            0                10               13   

       Transaction_Month  Transaction_Year  
0                     11              2018  
1                     11              2018  
2                     11              2018  
3                     11              2018  
4                     11              2018  
...                  ...               ...  
95657                  2              2019  
95658                  2              2019  
95659                  2              2019  
95660                  2              2019  
95661                  2              2019  

[95662 rows x 15 columns]